### PR TITLE
Do not include compiler-generated names in expression names

### PR DESCRIPTION
**Pull request by dougbu**
- #117
- see also aspnet/Mvc@a045324d3a and PR aspnet/Mvc#3027, the fix for aspnet/Mvc#2890

nits: copy some improved comments from aspnet/Mvc@a045324d3a

### CODE REVIEW

File: src/System.Web.Mvc/ExpressionHelper.cs [hunk 1]
```diff
@@ -59,7 +60,18 @@ public static string GetExpressionText(LambdaExpression expression)
                 else if (part.NodeType == ExpressionType.MemberAccess)
                 {
                     MemberExpression memberExpressionPart = (MemberExpression)part;
-                    nameParts.Push("." + memberExpressionPart.Member.Name);
+                    var name = memberExpressionPart.Member.Name;
+
+                    // If identifier contains "__", it is "reserved for use by the implementation" and likely compiler-
+                    // or Razor-generated e.g. the name of a field in a delegate's generated class.
+                    if (name.Contains("__"))
```
Comment (pranavkm) [inline src/System.Web.Mvc/ExpressionHelper.cs]: Would we need a compat switch to allow this? Seems like a super corner case though.

Comment (pranavkm) [reply src/System.Web.Mvc/ExpressionHelper.cs]: Presumably Razor doesn't stick a `CompilerGeneratedAttribute` or something along those lines on the member that we could use to infer this?

Comment (dougbu) [reply src/System.Web.Mvc/ExpressionHelper.cs]: We're already dealing with an infrequent case. Imaging a scenario where this fix _causes_ a problem is pretty extreme. I don't think the slight change is worth a compat switch. But, I'll ask around&hellip;

Roslyn does apply `[CompilerGenerated]` to its generated classes but not to fields in those classes. So, that attribute doesn't help distinguish `<>c__DisplayClass3_0.collection` (which is fine) from `<>c__DisplayClass5_1,CS$<>8__locals1` (which isn't). and `__` is what docs say. (See the last paragraph in [this section](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#identifiers) of the language spec for example.)

**Comment by pranavkm** [comment]
:shipit:

**Comment by dougbu** [comment]
b32ab42712

**Pull request merged**